cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/profile/ProfileDataset.java

```diff
@@ -199,19 +212,77 @@ private void changeProfileStatus(ProfileId profileId,
   }
 
+  /**
+   * Get assignments with the profile.
+   *
+   * @param profileId the profile id
+   * @throws NotFoundException if the profile is not found
+   * @return a set of entity ids which contains the entities that are assigned with the profile
+   */
+  public Set<EntityId> getProfileAssignment(ProfileId profileId) throws NotFoundException {
+    MDSKey rowKey = getRowKey(profileId);
+    Profile profile = table.get(rowKey, Profile.class);
+    if (profile == null) {
+      throw new NotFoundException(profileId);
+    }
+
+    return Collections.unmodifiableSet(new HashSet<>(table.list(rowKey, INDEX_COLUMN, EntityId.class)));
+  }
+
+  /**
+   * Add an assignment to the profile. Assignment can only be added if the profile is ENABLED
+   *
+   * @param profileId the profile id
+   * @param entityId the entity to add to the assgiment
+   */
+  public void addProfileAssignment(ProfileId profileId,
+                                   EntityId entityId) throws ProfileConflictException, NotFoundException {
+    Profile profile = table.get(getRowKey(profileId), Profile.class);
+    if (profile == null) {
+      throw new NotFoundException(profileId);
+    }
+    if (profile.getStatus() == ProfileStatus.DISABLED) {
+      throw new ProfileConflictException(
+        String.format("Profile %s is DISABLED. So no entity can be assigned to it", profileId.getProfile()), profileId);
+    }
+    table.write(getEntityIndexRowKey(profileId, entityId), INDEX_COLUMN, entityId);
+  }
+
+  /**
+   * Remove an assignment from the profile.
+   *
+   * @param profileId the profile id
+   * @param entityId the entity to remove from the association
+   * @throws NotFoundException if the profile is not found
+   */
+  public void removeProfileAssignment(ProfileId profileId, EntityId entityId) throws NotFoundException {
+    Profile profile = table.get(getRowKey(profileId), Profile.class);
+    if (profile == null) {
+      throw new NotFoundException(profileId);
+    }
+    table.delete(getEntityIndexRowKey(profileId, entityId));
+  }
+
   private MDSKey getRowKey(NamespaceId namespaceId) {
-    return getRowKey(namespaceId, null);
+    return getRowKey(namespaceId, null, null);
   }
 
   private MDSKey getRowKey(ProfileId profileId) {
-    return getRowKey(profileId.getNamespaceId(), profileId.getProfile());
+    return getRowKey(profileId.getNamespaceId(), profileId.getProfile(), null);
+  }
+
+  private MDSKey getEntityIndexRowKey(ProfileId profileId, EntityId entityId) {
+    return getRowKey(profileId.getNamespaceId(), profileId.getProfile(), entityId);
   }
 
-  private MDSKey getRowKey(NamespaceId namespaceId, @Nullable String profileName) {
+  private MDSKey getRowKey(NamespaceId namespaceId, @Nullable String profileName, @Nullable EntityId entityId) {
     MDSKey.Builder builder = new MDSKey.Builder().add(PROFILE_PREFIX).add(namespaceId.getEntityName());
     if (profileName != null) {
       builder.add(profileName);
     }
+    if (entityId != null) {
```

Comment:
this means there are additional rows that will be scanned for every profile. What happens when you list the profiles?  Also, if you are writing a row per entity, that means you don't actually need to write to a different column. I think you can keep this rowkey, but have a different prefix for it. Then you don't need all the changes to MetadataDataset either.